src/bms/parse/token_processor/random.rs

```diff
@@ -574,24 +579,41 @@ impl<R: Rng, N: TokenProcessor> RandomTokenProcessor<R, N> {
 impl<R: Rng, N: TokenProcessor> TokenProcessor for RandomTokenProcessor<R, N> {
     type Output = <N as TokenProcessor>::Output;
 
-    fn process<P: Prompter>(
+    fn process<'a, 't, P: Prompter>(
         &self,
-        input: &mut &[&TokenWithRange<'_>],
-        prompter: &P,
-    ) -> TokenProcessorResult<Self::Output> {
-        let mut activated = vec![];
-        all_tokens_with_range(input, prompter, |token| {
-            let res = match token.content() {
-                Token::Header { name, args } => self.on_header(name.as_ref(), args.as_ref())?,
+        ctx: &mut ProcessContext<'a, 't, P>,
+    ) -> Result<Self::Output, crate::bms::parse::ParseErrorWithRange> {
+        let mut activated: Vec<&'a TokenWithRange<'t>> = Vec::new();
+        ctx.all_tokens(|token, _prompter| {
+            let warn = match token.content() {
+                Token::Header { name, args } => self
+                    .on_header(name.as_ref(), args.as_ref())
+                    .map(|ow| ow.map(|w| w.into_wrapper(token)))?,
                 Token::Message { .. } => None,
-                Token::NotACommand(line) => self.on_comment(line)?,
+                Token::NotACommand(line) => self
+                    .on_comment(line)
+                    .map(|ow| ow.map(|w| w.into_wrapper(token)))?,
             };
             if self.is_activated() {
                 activated.push(token);
             }
-            Ok(res)
+            Ok(warn)
         })?;
-        self.next.process(&mut &activated[..], prompter)
+
+        // Process activated tokens with the next processor using a temporary context.
+        let mut tmp = &activated[..];
+        let mut view_ctx = ProcessContext {
+            input: &mut tmp,
+            prompter: ctx.prompter(),
+            reported: Vec::new(),
+        };
+        let out = self.next.process(&mut view_ctx)?;
+        // Collect nested warnings locally first, then drop the borrowed context.
+        let nested_reported = core::mem::take(&mut view_ctx.reported);
+        drop(view_ctx);
+        ctx.reported.extend(nested_reported);
```

Comment:
I want this recovery and merging warnings to be done with the checkpoint mechanism. The same goes for `impl TokenProcessor for SequentialProcessor`.